flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -391,23 +416,33 @@ public void onContainersAllocated(List<Container> containers) {
 
 					workerNodeMap.put(resourceId, new YarnWorkerNode(container));
 
-					try {
-						// Context information used to start a TaskExecutor Java process
-						ContainerLaunchContext taskExecutorLaunchContext = createTaskExecutorLaunchContext(
-							container.getResource(),
-							containerIdStr,
-							container.getNodeId().getHost());
-
-						nodeManagerClient.startContainer(container, taskExecutorLaunchContext);
-					} catch (Throwable t) {
-						log.error("Could not start TaskManager in container {}.", container.getId(), t);
-
-						// release the failed container
-						workerNodeMap.remove(resourceId);
-						resourceManagerClient.releaseAssignedContainer(container.getId());
-						// and ask for a new one
-						requestYarnContainerIfRequired();
-					}
+					startContainerExecutor.execute(new Runnable() {
+						@Override
+						public void run() {
+							if (workerNodeMap.get(resourceId) == null) {
+								log.info("Skip launching container {}, container doesn't exist in container worker map",
+									containerIdStr);
+								return;
+							}
+							try {
+								// Context information used to start a TaskExecutor Java process
+								ContainerLaunchContext taskExecutorLaunchContext = createTaskExecutorLaunchContext(
+									container.getResource(),
+									containerIdStr,
+									container.getNodeId().getHost());
+
+								nodeManagerClient.startContainer(container, taskExecutorLaunchContext);
+							} catch (Throwable t) {
+								log.error("Could not start TaskManager in container {}.", container.getId(), t);
+
+								// release the failed container
+								workerNodeMap.remove(resourceId);
+								resourceManagerClient.releaseAssignedContainer(container.getId());
+								// and ask for a new one
+								requestYarnContainerIfRequired();
```

Comment:
This method must not be called from outside of the `RpcEndpoint's` main thread because it mutates the internal state of the `ResourceManager`.